File: worker/src/cardinality.ts
/**
 * Cardinality Governance Module
 */

import { Config } from './config';
import { Env } from './types';

export class CardinalityService {
  private config: Config;
  private env: Env;

  constructor(config: Config, env: Env) {
    this.config = config;
    this.env = env;
  }

  /**
   * Apply cardinality governance to a dimension value
   */
  async applyGovernance(dimension: string, value: string): Promise<string | null> {
    const limit = this.config.getCardinalityLimit(dimension);

    if (!limit) {
      // No limit defined, allow the value
      return value;
    }

    switch (limit.action) {
      case 'allow':
        return await this.handleAllow(dimension, value, limit.max_cardinality);

      case 'bucket':
        return this.handleBucket(dimension, value, limit.bucket_size || 1000);

      case 'hash':
        return this.handleHash(value);

      case 'drop':
        return null;

      default:
        return value;
    }
  }

  /**
   * Handle 'allow' action - track cardinality and allow if under limit
   */
  private async handleAllow(dimension: string, value: string, maxCardinality: number): Promise<string | null> {
    const key = `cardinality:${dimension}`;

    try {
      // Get current cardinality set from KV
      const existingJson = await this.env.CARDINALITY_KV.get(key);
      const existingSet: Set<string> = existingJson
        ? new Set(JSON.parse(existingJson))
        : new Set();

      // Check if value already exists
      if (existingSet.has(value)) {
        return value; // Already tracked, allow
      }

      // Check if adding this value would exceed limit
      if (existingSet.size >= maxCardinality) {
        console.warn(`Cardinality limit reached for ${dimension}: ${existingSet.size}/${maxCardinality}`);
        return this.handleHash(value); // Fall back to hash
      }

      // Add to set and save back to KV
      existingSet.add(value);
      await this.env.CARDINALITY_KV.put(
        key,
        JSON.stringify(Array.from(existingSet)),
        { expirationTtl: 86400 } // 24 hours TTL
      );

      return value;
    } catch (error) {
      console.error(`Error tracking cardinality for ${dimension}:`, error);
      return value; // Allow on error
    }
  }

  /**
   * Handle 'bucket' action - bucket values into ranges
   */
  private handleBucket(dimension: string, value: string, bucketSize: number): string {
    // For numeric values, bucket into ranges
    const numValue = parseFloat(value);
    if (!isNaN(numValue)) {
      const bucketIndex = Math.floor(numValue / bucketSize);
      const bucketStart = bucketIndex * bucketSize;
      const bucketEnd = bucketStart + bucketSize;
      return `${bucketStart}-${bucketEnd}`;
    }

    // For string values, use hash bucketing
    return this.hashBucket(value, bucketSize);
  }

  /**
   * Handle 'hash' action - hash the value
   */
  private handleHash(value: string): string {
    return this.simpleHash(value);
  }

  /**
   * Hash value into buckets
   */
  private hashBucket(value: string, bucketSize: number): string {
    const hash = this.simpleHashNumber(value);
    const bucket = hash % bucketSize;
    return `bucket_${bucket}`;
  }

  /**
   * Simple hash function for strings (returns hash string)
   */
  private simpleHash(str: string): string {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32-bit integer
    }
    return Math.abs(hash).toString(16).padStart(8, '0');
  }

  /**
   * Simple hash function for strings (returns number)
   */
  private simpleHashNumber(str: string): number {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash;
    }
    return Math.abs(hash);
  }

  /**
   * Apply governance to all dimensions in a label set
   */
  async applyGovernanceToLabels(labels: Record<string, string>): Promise<Record<string, string>> {
    const governed: Record<string, string> = {};

    for (const [key, value] of Object.entries(labels)) {
      const governedValue = await this.applyGovernance(key, value);
      if (governedValue !== null) {
        governed[key] = governedValue;
      }
    }

    return governed;
  }

  /**
   * Get current cardinality stats for a dimension
   */
  async getCardinalityStats(dimension: string): Promise<{ count: number; values: string[] }> {
    const key = `cardinality:${dimension}`;

    try {
      const existingJson = await this.env.CARDINALITY_KV.get(key);
      if (!existingJson) {
        return { count: 0, values: [] };
      }

      const values = JSON.parse(existingJson) as string[];
      return { count: values.length, values };
    } catch (error) {
      console.error(`Error getting cardinality stats for ${dimension}:`, error);
      return { count: 0, values: [] };
    }
  }

  /**
   * Reset cardinality tracking for a dimension
   */
  async resetCardinality(dimension: string): Promise<void> {
    const key = `cardinality:${dimension}`;
    await this.env.CARDINALITY_KV.delete(key);
  }
}

File: worker_v2/config/cardinality.go
package config

import (
	"encoding/json"
	"fmt"
	"hash/crc32"
	"math"
	"strconv"
	"time"

	"go.uber.org/zap"
	datastructure "openqoe.dev/worker_v2/data_structure"
)

type Action string

const (
	Allow  Action = "allow"
	Bucket Action = "bucket"
	Hash   Action = "hash"
	Drop   Action = "drop"
)

type CardinalityLimit struct {
	Max_Cardinality float64 `json:"max_cardinality"`
	Action          Action  `json:"action"`
	Bucket_Size     int     `json:"bucket_size"`
}

type CardinalityStat struct {
	Count  int
	Values []string
}

type CardinalityConfig struct {
	Limits map[string]CardinalityLimit
}

type CardinalityService struct {
	env    *Env
	config *Config
	logger *zap.Logger
}

func NewCardinalityService(env *Env, config *Config, parent_logger *zap.Logger) *CardinalityService {
	logger := parent_logger.With(zap.String("sub-component", "cardinality_service"))
	return &CardinalityService{
		env:    env,
		config: config,
		logger: logger,
	}
}

func (cs *CardinalityService) applyGovernance(dimension string, value string) string {
	limit := cs.config.GetCardinalityLimit(dimension)
	res_chan := make(chan string, 1)
	switch limit.Action {
	case Allow:
		go cs.handleAllow(dimension, value, limit.Max_Cardinality, res_chan)
	case Bucket:
		go cs.handleBucket(value, limit.Bucket_Size, res_chan)
	case Hash:
		go cs.handleHash(value, res_chan)
	case Drop:
		go func() {
			defer close(res_chan)
			res_chan <- ""
		}()
	default:
		go func() {
			defer close(res_chan)
			res_chan <- value
		}()
	}
	return <-res_chan
}

// Handle 'allow' action - track cardinality and allow if under limit
func (cs *CardinalityService) handleAllow(dimension string, value string, max_cardinality float64, res_chan chan<- string) {
	defer close(res_chan)
	const key = "cardinality:" // Redis key prefix
	// Get current cardinality set from K
	existing_json, err := cs.config.Redis_client.GetValue(key + dimension)
	if err != nil {
		cs.logger.Error("Failed to get existing cardinality data", zap.Error(err))
	}

	existing_set := datastructure.NewSet[string]()
	if existing_json != "" {
		var items []string
		if err := json.Unmarshal([]byte(existing_json), &items); err != nil {
			cs.logger.Error("Failed to unmarshal existing cardinality data", zap.Error(err))
		} else {
			existing_set = datastructure.NewSet(items...)
		}
	}
	// Check if value already exists
	if existing_set.Contains(value) {
		res_chan <- value
		return
	}

	// Check if adding this value would exceed limit
	if existing_set.Size() >= int(max_cardinality) {
		cs.logger.Warn("cardinality limit reached", zap.String("dimension", dimension), zap.Int("limit", existing_set.Size()/int(max_cardinality)))
		res_chan <- cs.handleHashString(value)
		return
	}

	// Add to set and save back to KV
	existing_set.Add(value)
	items := existing_set.Items()
	json_data, err := json.Marshal(items)
	if err != nil {
		cs.logger.Error("Failed to marshal cardinality data", zap.Error(err))
		return
	}
	err = cs.config.Redis_client.SetValueWithTTL(key+dimension, string(json_data), 86400*time.Second)
	if err != nil {
		cs.logger.Error("Failed to set cardinality data with TTL", zap.Error(err))
	}
	res_chan <- value
}

func (cs *CardinalityService) handleBucket(value string, bucket_size int, res_chan chan<- string) {
	defer close(res_chan)
	num_value, err := strconv.ParseFloat(value, 64)

	if err != nil {
		res_chan <- cs.handleHashBucket(value, bucket_size)
	}

	bucket_index := math.Floor(num_value / float64(bucket_size))
	bucket_start := bucket_index * float64(bucket_size)
	bucket_end := bucket_start + float64(bucket_size)
	res_chan <- fmt.Sprintf("%d-%d", int(bucket_start), int(bucket_end))

}
func (cs *CardinalityService) handleHash(value string, res_chan chan<- string) {
	defer close(res_chan)
	res_chan <- cs.handleHashString(value)
}

func (cs *CardinalityService) handleHashString(value string) string {
	hash := crc32.ChecksumIEEE([]byte(value))
	return fmt.Sprintf("%08x", hash)
}

func (cs *CardinalityService) handleHashBucket(value string, bucket_size int) string {
	hash := cs.handleHashNumber(value)
	bucket := int(hash) % bucket_size
	return fmt.Sprintf("bucket_%d", bucket)
}

func (cs *CardinalityService) handleHashNumber(value string) uint32 {
	return crc32.ChecksumIEEE([]byte(value))
}

func (cs *CardinalityService) ApplyGovernanceToLabels(labels map[string]string) map[string]string {
	governed := make(map[string]string)
	for key, value := range labels {
		governed_value := cs.applyGovernance(key, value)
		if governed_value != "" {
			governed[key] = governed_value
		}
	}
	return governed
}

func (cs *CardinalityService) GetCardinalityStats(dimension string) *CardinalityStat {
	const key = "cardinality:"
	existing_json, err := cs.config.Redis_client.GetValue(key + dimension)
	if err != nil || existing_json == "" {
		cs.logger.Error("Failed to get existing cardinality data", zap.Error(err))
		return &CardinalityStat{Count: 0, Values: []string{}}
	}
	var values []string
	if err := json.Unmarshal([]byte(existing_json), &values); err != nil {
		cs.logger.Error("Error getting cardinality stats", zap.Error(err), zap.String("dimension", dimension))
		return &CardinalityStat{Count: 0, Values: []string{}}
	}
	return &CardinalityStat{Count: len(values), Values: values}
}

func (cs *CardinalityService) resetCardinality(dimension string, res_chan chan<- bool) {
	defer close(res_chan)
	const key = "cardinality:"
	err := cs.config.Redis_client.Delete(key + dimension)
	if err != nil {
		cs.logger.Error("Error resetting cardinality", zap.Error(err), zap.String("dimension", dimension))
		res_chan <- false
		return
	}
	res_chan <- true
}
